src/v/cluster/cluster_epoch_service.cc

```diff
@@ -0,0 +1,367 @@
+// Maximum amount of time to cache the same epoch before we block on the update.
+constexpr ss::lowres_clock::duration max_same_epoch_cache_duration = 60min;
+
+} // namespace
+
+// raft0_state is a service that runs on shard0 and is responsible for
+// managing the current epoch. The current epoch is a frozen point in time of
+// the committed offset of the controller log. We need to use the controller log
+// so that we can correlate partition creation with the current epoch.
+//
+// We periodically "freeze" and refresh the epoch so that we control the epoch
+// change and don't have to worry about random cluster operations causing more
+// frequent epoch changes (which have the potential for distrupting L0
+// operations in cloud topics).
+class cluster_epoch_service::raft0_state {
+public:
+    raft0_state(
+      ss::lw_shared_ptr<raft::consensus> raft0,
+      ss::sharded<controller_stm>& stm,
+      ss::sharded<raft::group_manager>& group_manager)
+      : _raft0(std::move(raft0))
+      , _stm(stm)
+      , _group_manager(group_manager)
+      , _queue([](const std::exception_ptr& err) {
+          vlog(clusterlog.error, "Error in cluster epoch service: {}", err);
+      })
+      , _leadership_notification_id(
+          group_manager.local().register_leadership_notification(
+            [this](
+              raft::group_id group_id,
+              model::term_id,
+              std::optional<model::node_id>) {
+                if (group_id != _raft0->group()) {
+                    return;
+                }
+                if (_raft0->is_leader()) {
+                    start_service_loop();
+                } else {
+                    stop_service_loop();
+                }
+            })) {}
+
+    // Shutdown this state, stopping the service loop if it's running.
+    //
+    // Must be called before destruction.
+    ss::future<> stop() {
+        _group_manager.local().unregister_leadership_notification(
+          _leadership_notification_id);
+        co_await _queue.shutdown();
+        // Now that the queue won't accept new tasks, we can safely stop the
+        // service loop if it's running.
+        _abort_source.request_abort();
+        co_await std::exchange(_loop, std::nullopt)
+          .or_else([]() { return std::make_optional(ss::now()); })
+          .value();
+    }
+
+    // Get the current epoch - only set if this node is the leader of raft0.
+    std::optional<model::offset> current_epoch() const {
+        return _current_epoch;
+    }
+
+private:
+    void start_service_loop() {
+        _queue.submit([this] {
+            _abort_source = {};
+            _loop = service_loop();
+            return ss::now();
+        });
+    }
+    void stop_service_loop() {
+        _queue.submit([this] {
+            _abort_source.request_abort();
+            return std::exchange(_loop, std::nullopt)
+              .or_else([]() { return std::make_optional(ss::now()); })
+              .value();
```

Comment:
handle exception and ignore / log?

Reply:
It **should** never throw, but I will add the logic anyways

Reply:
done